Additional data visualizations

gaborposz opened the pull request:
Hello,
It's me again! :)

I added some additional visualization features to the `InspectObjectToolSystem`, the ones which I was missing the most.
Here comes the details:

- I adapted the `ComponentDataRenderer` to support hovering over Component items (i.e. fields of Components) as well, not only over Components. 
  - Renamed the `HoverData.Type` to `ComponentType` (as it will always contain the type of the Component, even if a child item was hovered) and added a new field for a `ComponentItem` reference.
  - Replaced the `out bool hovered` argument of the `Render` method to `out IInspectableObject hoveredObject`. This lets me get the hovered ComponentItems, not only the type of the parent Component.
  - If there was a hoovered ComponentItem down the tree, then that is returned by the `Render` method at the top of the recursive call, and then it is saved in the `lastHovered.ComponentItem` field. The `lastHovered.ComponentType` behaves as before, so if there is no visualization supported for the ComponentItem, then the `InspectObjectToolSystem` can fall back to the visualization of the (parent) Component (see below).
- I modified the `ComponentDataRenderer.Render` method to also display the type of the Component items to make it more obvious for the user what is visualized, and what is not.
- I added a logic to `InspectObjectToolSystem.OnUpdate` that implements the following logic: If the `RenderByManagedType` method could visualize the ComponentItem (i.e. its type was supported), then the `OnUpdate` stops, but if it was not supported, then it falls back to the rendering of the Component (like before).
- And finally, the list of the new visualizations:
  - **Bezier4x3**: If a Component has an item (i.e. field) with the a bezier curve then it is drawn to the screen. It lets the user check the sub geometry of Components, and also provides some level of visualization for not supported Components.
  
  ![image](https://github.com/user-attachments/assets/93649fd8-e167-47e7-a18d-fde31d42d0cf)

  - **Bounds3**: Similar to bezier curve rendering, it visualizes bounding boxes. Unfortunately the used `OverlayRenderSystem` does not support drawing vertical lines, so only the top and the bottom of the box is drawn.
  
  ![image](https://github.com/user-attachments/assets/27515285-fed8-4ba3-b70a-fa30b6c65c3f)

  - **float3**: Assuming that it is a 3D location, a red cross is drawn on the map.
  
  ![image](https://github.com/user-attachments/assets/0af3a909-db09-41b6-bef3-47082f684ff2)

  - **Game.Net.Segment**: This is already a regular Component, drawn by its bezier curves. (It is a good example for testing that visualizations are now supported both on the Item and Component levels.)
  
  ![image](https://github.com/user-attachments/assets/e6b53dd4-4ea6-445e-b93f-925681e49b4a)

  - **Game.Pathfind.PathElement**: It visualizes the path (planned by the game's pathfind logic). It also supports visualization of transit lines, which are not entirely part of the PathElement buffer, but only the start and end Waypoints (i.e. the start and end transit stops). The code retrieves PathElements from the relevant segments of the transit line, and draws them recursively by `RenderPathElements`.
  
  ![image](https://github.com/user-attachments/assets/2fa8030b-1282-4b3d-93ac-b579515ebb51)

  - **Game.Vehicles.CarNavigationLane, .TrainNavigationLane, .WatercraftNavigationLane, .AircraftNavigationLane**: These are similar to the PathElement buffer, but they contain the "short term" route of the creature with additional details, like concrete sublanes.
  
  ![image](https://github.com/user-attachments/assets/6c033ab6-b9e8-4d1e-90a3-ba073cf0f5dd)


Let me know your thoughts!

I tried to test my changes as much as possible, but there is always a chance that something goes south in specific circumstances (like with my previous PR 😄). If you find something, or get negative feedback, than let me know, I'll try to respond and create a fix as soon as I can.

gaborposz committed:
- Support for returning the "hovered item" on all levels (not only on Component level)
- Data visualization for Bezier4x3, Bounds3 and Segment

gaborposz committed:
Data visualization for
- float3
- PathElement (buffer)
- NavigationLanes (buffer)

+ Hovered object fixes

gaborposz committed:
Fix of sporadic exception:
                    NullReferenceException: Object reference not set to an instance of an object
                      at Unity.Collections.LowLevel.Unsafe.UnsafeUtility.ReadArrayElement[T] (System.Void* source, System.Int32 index) [0x00000] in <b4e80bd296bc48ba89d931486f299d70>:0
                      at Unity.Collections.NativeArray`1+Enumerator[T].MoveNext () [0x00027] in <b4e80bd296bc48ba89d931486f299d70>:0
                      at SceneExplorer.System.InspectObjectToolSystem.RenderPathElements (Unity.Entities.Entity entityWithPathElements, Game.Rendering.OverlayRenderSystem+Buffer buffer) [0x0004a] in C:\Users\gabor\Source\Repos\SceneExplorer\SceneExplorer\System\InspectObjectToolSystem.cs:621

gaborposz committed:
Not needed usings were removed.

gaborposz committed:
Support for rendering only the RouteSegments between the transit line waypoints. Duplicate rendering of the transit line was also eliminated.

gaborposz committed:
Bumped the version number to 0.2.10 and added ChangeLog.

gaborposz commented:
@krzychu124 could you have a look on this?

krzychu124 commented:
@gaborposz really sorry, I was busy recently. I'll take a look tomorrow

GaborPosz-EF commented:
@krzychu124 no worries, this is a hobby project, no commitments :)

krzychu124 commented:
Hmm, I could merge your changes but replace rendering bounding box, bezier, segment and node with `GizmoBatcher`
Here highlighting `SubObject` category, rendering bounding boxes of all objects.

![image](https://github.com/user-attachments/assets/bca675cb-a229-451d-aa82-9da2f74878c2)

What do you think?

gaborposz commented:
You mean to use `GizmoBatcher` to draw these types?
If `GizmoBatcher` can draw vertical lines (`OverlayRenderSystem` apparently can't), then it's a great idea! :)
I'll look into it.

krzychu124 commented:
> You mean to use `GizmoBatcher` to draw these types? If `GizmoBatcher` can draw vertical lines (`OverlayRenderSystem` apparently can't), then it's a great idea! :) I'll look into it.

Check out `Gizmos` tab in vanilla dev menu if you haven't already. Everything there is done with gizmo batcher. The only limit is line width cannot be changed, it's always 1 pixel so visible at any distance.

I can push my changes to `main` branch (only few tweaks + job for generating gizmos from a list of entities) then you could adjust your features/improvements to use it too.

gaborposz committed:
- Using GizmoBatcher instead of OverlayRenderer
- Inifinite loop bug was fixed in PathElement rendering
- Warning was removed because it was logged too many times (despite that the visualization looked ok)

gaborposz commented:
In the meantime I found some examples in [Traffic](https://github.com/krzychu124/Traffic/blob/5668d89c4013fd93d0a12eebc272b361e71d6059/Code/Debug/LaneConnectorDebugSystem.cs#L79) and adapted all the new renderings that I added. Looks good!
I also found an infinite loop during testing, and hopefully fixed it. 🤞
The ElementPath rendering (`RenderPath`) is relative slow. Do you think that I should also push that out into a job?

gaborposz committed:
Path rendering performance optimizations:
- The whole rendering is moved out of the main thread into a job
- Burts compilation was enabled for the PathRenderingJob
- Bezier rendering segment count was limited to 6 (instead of 16)

gaborposz commented:
I moved the PathRendering to a burst compiled job. The performance is much better now.
Let me know if I should change anything else.

krzychu124 commented on SceneExplorer/System/InspectObjectToolSystem.cs:
```diff
@@ -37,6 +43,105 @@ public struct InspectedObject : IComponentData
+        [BurstCompile]
+        private struct PathRendererJob : IJob
+        {
+            [ReadOnly] public GizmoBatcher gizmoBatcher;
```
`gizmoBatcher` **is not** read only here. It's ReadWrite, so default (no attribute)

krzychu124 commented on SceneExplorer/System/InspectObjectToolSystem.cs:
```diff
@@ -37,6 +43,105 @@ public struct InspectedObject : IComponentData
+            [ReadOnly] public ComponentLookup<CullingInfo> cullingInfoCompontentLookup;
+                
+            private bool startWaypointSet;
+            private Game.Routes.Waypoint startWayPoint;
```
Both could be passed by ref to the method, instead of using private field. Up to you.

gaborposz replied:
I refactored it to a `ref Game.Routes.Waypoint?` argument.
And actually this change turned out to be useful, because it highlighted the fact that in the recursive call this variable could be overwritten, so I introduced another another variable for the inner loop (`innerStartWayPoint`). This is a theoretical scenario though, because transit routes does not contain WayPoints.

krzychu124 commented on SceneExplorer/System/InspectObjectToolSystem.cs:
```diff
@@ -468,6 +663,39 @@ private bool RenderByManagedType(Type managedType, ComponentDataRenderer.HoverDa
+            {
+                gizmoBatcher = _gizmosSystem.GetGizmosBatcher(out JobHandle dependencies),
+                pathEntities = pathEntitiesAndDeltas,
+                curveCompontentLookup = GetComponentLookup<Curve>(true),
```
Please use `SystemAPI`

krzychu124 commented on SceneExplorer/System/InspectObjectToolSystem.cs:
```diff
@@ -468,6 +663,39 @@ private bool RenderByManagedType(Type managedType, ComponentDataRenderer.HoverDa
+                takeoffLocationCompontentLookup = GetComponentLookup<Game.Routes.TakeoffLocation>(true),
+                spawnLocationCompontentLookup = GetComponentLookup<Game.Objects.SpawnLocation>(true),
+                cullingInfoCompontentLookup = GetComponentLookup<CullingInfo>(true),
+            }.Schedule();
```
I think you forgot about passing `GizmoBatcher` dependency to `Schedule` (and combine it with this system dependency (`this.Dependency`) or dependency from the outside of method if expected to run multiple times). It might crash the game at worst case if some other mod will use gizmo batcher too.

gaborposz replied:
Sorry, I'm still trying to comprehend the Unity Job System's dependency concept. Please check if I changed it correctly.

krzychu124 replied:
Hmm, there's a problem, you should combine `inputDeps` param with `dependencies` (I didn't notice you are passing it as argument when I was writing the message).
It's not so hard to figure out the job system. The most important thing to keep in mind is that `Schedule()` never starts the job immediately. It's only preparing data and returns the handle (kind of a pointer to the job). When you want to force job completion there are basically two options: 
- call `.Complete()` on the returned handle (it will freeze main thread and wait for completion)
- pass the handle to other `Schedule()`

Then based on native container access type (readwrite, write/read-only) of job component buffers, lookups and handles unity will try to run as many jobs as possible (different type of jobs (*might) run in parallel, while the same type (*might) run parallel only if you call `ScheduleParallel`)

*sequential processing will be forced for jobs with potential race conditions - reading and writing to the same containers

gaborposz replied:
I understand the basics, it is very similar to the principles of TPL, but I get confused with the `Dependency` property.
- Why do we need to add every job's handle to it? To make sure that the system will not start running again (e.g. in the next frame) before the jobs get finished?
- Why do we need to make every new job (via the `Dependency` property) depend on the previous job? They might be independent and run in parallel, but if we make them sequential...
- Why does the `GizmoSystem.GetGizmosBatcher` return another JobHandle (`out JobHandle dependencies`)? Is it the handle of the previous "gizmo drawing" job, so that we can schedule our job after that?
- Do we pass our job's handle to `GizmoSystem.AddGizmosBatcherWriter` to make sure that the gizmo drawing is not running before we "drawn" (i.e. added) all of our gizmos?

Anyway, I replaced `Dependency` to `inputDeps`. If my assumptions above were right, then it should be good now.

krzychu124 replied:
> Why do we need to add every job's handle to it? To make sure that the system will not start running again (e.g. in the next frame) before the jobs get finished?

Nope, IMO, it's because you can have the system as dependency, "this.Dependencies" is the system dependency handle.
> Why do we need to make every new job (via the Dependency property) depend on the previous job? They might be independent and run in parallel, but if we make them sequential...

That's not a requirement. Again, `Dependency` is the system dependency, can be combined from outside and you are not forced to use it, but it's good practice to set it somewhere in the `OnUpdate` (see [SystemState.Dependency](https://docs.unity3d.com/Packages/com.unity.entities@1.0/api/Unity.Entities.SystemState.Dependency.html) ) At least from what I understand.
> Why does the GizmoSystem.GetGizmosBatcher return another JobHandle (out JobHandle dependencies)? Is it the handle of the previous "gizmo drawing" job, so that we can schedule our job after that?

It's the `GizmoSystem` `Dependency` handle under the hood, it helps job scheduler decide if yours and other jobs can run in parallel when you expect it or force sequential based on read-write access. 
There's a difference between running the same job in parallel (using multiple workers) vs concurrent (running different jobs from different systems at the same time). System dependency helps job scheduler with both.
> Do we pass our job's handle to GizmoSystem.AddGizmosBatcherWriter to make sure that the gizmo drawing is not running before we "drawn" (i.e. added) all of our gizmos?

Yes, basically to delay execution. It's a handle that will be used by `GizmoSystem` to force job completion of all combined handles. Depends on your needs you might use it once (e.g.: passing final job handle when chaining multiple jobs or system dependency), or multiple times (when you allow/want concurrent work).

krzychu124 commented on SceneExplorer/System/InspectObjectToolSystem.cs:
```diff
@@ -392,6 +512,81 @@ private bool RequireManualHighlight(Entity entity, out ComponentType componentTy
+            {
+                var gizmoBatcher = _gizmosSystem.GetGizmosBatcher(out JobHandle dependencies);
+                deps = JobHandle.CombineDependencies(inputDeps, dependencies);
+                RenderCurve(hovered, gizmoBatcher);
```
😬 Not good, IMO, `GizmosBatcher` shouldn't be used directly (without respecting dependency). It works ok, only because this system doesn't schedule any other dependent work (which is not the case e.g. in _Traffic_ mod).
It suppose to work like this:
- take ref to batcher and the "write" dependency handle
- schedule job which is dependent on "write" dependency (will run only when allowed to - previous (unknown) dependencies are completed)
- register returned job handle as batcher writer (simply combines itself with previous gizmo system dependencies)

When the game is using batcher, it will clear it first and force completion of all jobs registered as writers. If you use batcher directly, there is a chance all draw requests you created will be ignored (cleared before execution), plus missing/incorrect use of dependency may lead to game crash or suboptimal job scheduling: 
- some jobs might stall, not that bad, it will just waste CPU time, so slow down rendering gizmos but only vanilla is using it at the moment AFAIK and only for debug visualization,
- jobs might actually run in parallel when they shouldn't, which might crash the game at worst case

gaborposz replied:
So if I understand it correctly, the problem is that I might draw into `GizmoBatcher` directly when another parallel job is trying to do the same at the same time. Honestly, I was just copying the pattern from `OverlayRenderSystem.GetBuffer`, and I don't understand why it was ok there, but not ok here?

Anyway, what would be the solution?
Pushing out all the drawings to jobs? It is possible, but seems to be quite a bit of overhead for a few draw calls.
Is there any other safe option?

krzychu124 replied:
It's not correct for `OverlayRenderSystem` either 😅 but like I mentioned it should work in the current state.
It's more about compatibility with other mods than with vanilla. This UI System doesn't offload or wait for dependencies so (almost) all code is running on main thread, so vanilla gizmos will wait.

gaborposz replied:
Alright, then I leave it as is, until we have a better idea :)

krzychu124 commented on SceneExplorer/System/InspectObjectToolSystem.cs:
```diff
@@ -468,6 +663,39 @@ private bool RenderByManagedType(Type managedType, ComponentDataRenderer.HoverDa
+            where T : unmanaged
+        {
+            JobHandle deps;
+            var pathEntitiesAndDeltas = new NativeArray<ValueTuple<Entity, float2>>(pathEntities.Capacity, Allocator.TempJob);
```
1. Memory leak. Add `pathEntitiesAndDeltas.Dispose(jobHandle);` after the job `Schedule()` call so it will be disposed when job is done.
2. Why not `Length` instead of `Capacity`? 
3. You can make it tiny bit faster when you use 3rd argument `NativeArrayOptions.UninitializedMemory` since it's overwriting every array value (clearing in advance is waste of CPU time in this case).

gaborposz replied:
Regarding point 1: Isn't the memory cleaned up anyway after the frame if the `TempJob` allocator is used?

krzychu124 replied:
It's almost never done automatically for `TempJob` and never for `Persistent` allocator. Only `Temp` does that. It's because the `TempJob` can live up to 4 frames, then you will see warnings about memory leak

gaborposz replied:
Ahh ok, I mixed `TempJob` with `Temp`. I got it now.

krzychu124 reviewed (changes requested):
I left few comments but in general I think you could try changing `PathRendererJob` to simply accept entity(ies) (or its archetypechunk) and figure out the rest inside the job (check type). 
It's a bit more advanced way of doing things but might help in other cases.
I can handle it if you want.

gaborposz committed:
Review finding fixes

gaborposz commented:
> I left few comments but in general I think you could try changing `PathRendererJob` to simply accept entity(ies) (or its archetypechunk) and figure out the rest inside the job (check type). It's a bit more advanced way of doing things but might help in other cases. I can handle it if you want.

I was considering it, but I don't know how to create a query for this use case.
The `PathRendererJob` is enumerating "path entities" (+ also intervals) from buffer components of a the actually selected entity. Moreover, the buffer components (and the way how the entities are extracted from them) are different in the different scenarios.
E.g.:
`DynamicBuffer<PathElement>` -> `pathElement.m_Target`, `pathElement.m_TargetDelta`
`DynamicBuffer<CarNavigationLane>` -> `pathElement.m_Lane`, `pathElement.m_CurvePosition`

Do you have any suggestions?


By the way, thank you for the findings, I'm still learning Unity (frankly, this was the first Unity Job that I ever created), and these comments help a lot!

krzychu124 commented:
> > I left few comments but in general I think you could try changing `PathRendererJob` to simply accept entity(ies) (or its archetypechunk) and figure out the rest inside the job (check type). It's a bit more advanced way of doing things but might help in other cases. I can handle it if you want.
> 
> I was considering it, but I don't know how to create a query for this use case. The `PathRendererJob` is enumerating "path entities" (+ also intervals) from buffer components of a the actually selected entity. Moreover, the buffer components (and the way how the entities are extracted from them) are different in the different scenarios. E.g.: `DynamicBuffer<PathElement>` -> `pathElement.m_Target`, `pathElement.m_TargetDelta` `DynamicBuffer<CarNavigationLane>` -> `pathElement.m_Lane`, `pathElement.m_CurvePosition`
> 
> Do you have any suggestions?
> 
> By the way, thank you for the findings, I'm still learning Unity (frankly, this was the first Unity Job that I ever created), and these comments help a lot!

Yeah, I know it's not possible (in current implementation) to generate query that could match only the selected entity, but there are few other options.
You can get whole `ArchetypeChunk` (struct of arrays of entity components) and inside the job (or outside) find the entity, then use `chunk.Has(ref someComponentTypeHandle)` to check if entity contains certain component then do stuff based on results (it will be faster than `EntityManager.HasComponent<>` because of burst).
Since entity doesn't have different path element buffers (either one or the other) you would have `if (Has(ref Xtype)){} else if(Has(ref YType)){ }` etc. which later could be extended with more variants.

I think you can leave it as is and I'll refactor it, since I need something similar and there is no need to have separate job for each type of object when the job could simply process array of entities (e.g.: in parallel)

gaborposz committed:
Fix of wrong dependency variable (at this point inputDeps might contain already more dependencies than Dependency)

gaborposz commented:
I tested the changes again, and it looks good to go.

krzychu124 reviewed (approved):
Looks ok, I'll add my changes too, before the release

krzychu124 commented:
Next time I'll try to create a new branch for upcoming version, so will be easier to keep track of changes and commit history will be cleaner since last commit is shown in the mod options with a link. 
I'd like it to be a commit collecting all changes done in the version, so e.g.: a PR merge `version-xyz` to `main`, where `version-xyz` will be PR target if you or someone else wants to contribute. 
Alternatively, I might create `dev` branch as default PR target and then push commits/merged PR's to `main` and build the version from it 🤔 
There's no need to update xml file, you can put changes in the PR description because in most cases I'll need to add something there too 😉

Anyways, @gaborposz  thanks for contribution 🙏

GaborPosz-EF commented:
You're welcome, thanks for your constructive comments!

If I can add my two cents, I would prefer a single `develop` branch.
This is also the branching strategy that we usually use at my company. It is pretty simple and flexible. You can merge to `main` whenever you want to release, and you can also do cherry picking when necessary (very rare). And you can also set up automation on `main`, e.g. for automatic versioning, deployment, or even for release note generation.

gaborposz commented:
You're welcome, thanks for your constructive comments!

If I can add my two cents, I would prefer a single `develop` branch.
This is also the branching strategy that we usually use at my company. It is pretty simple and flexible. You can merge to `main` whenever you want to release, and you can also do cherry picking when necessary (very rare). And you can also set up automation on `main`, e.g. for automatic versioning, deployment, or even for release note generation.
